mmdet/models/roi_heads/sparse_roi_head.py

```diff
@@ -139,6 +144,37 @@ def _bbox_forward(self, stage, x, rois, object_feats, img_metas):
 
         return bbox_results
 
+    def _mask_forward(self, stage, x, rois, attn_feats):
+        mask_roi_extractor = self.mask_roi_extractor[stage]
+        mask_head = self.mask_head[stage]
+        mask_feats = mask_roi_extractor(x[:mask_roi_extractor.num_inputs],
+                                        rois)
+        # do not support caffe_c4 model anymore
+        mask_pred = mask_head(mask_feats, attn_feats)
+
+        mask_results = dict(mask_pred=mask_pred)
+        return mask_results
+
+    def _mask_forward_train(self, stage, x, attn_feats, sampling_results, gt_masks, rcnn_train_cfg):
+
+        if sum([len(gt_mask) for gt_mask in gt_masks])==0:
+            print('Ground Truth Not Found!')
+            loss_mask = sum([_.sum() for _ in self.mask_head[stage].parameters()]) * 0.
+            return dict(loss_mask=loss_mask)
+        pos_rois = bbox2roi([res.pos_bboxes for res in sampling_results])
+        attn_feats = torch.cat([feats[res.pos_inds] for (feats, res) in zip(attn_feats, sampling_results)])
+        mask_results = self._mask_forward(stage, x, pos_rois, attn_feats)
+
+        mask_targets = self.mask_head[stage].get_targets(
+            sampling_results, gt_masks, rcnn_train_cfg)
+
+        pos_labels = torch.cat([res.pos_gt_labels for res in sampling_results])
+
+        loss_mask = self.mask_head[stage].loss(mask_results['mask_pred'],
+                                               mask_targets, pos_labels)
+        mask_results.update(loss_mask)
+        return mask_results
+
     def forward_train(self,
                       x,
```

Comment:
Don't use the name x

Reply:
This code is a part of original SparseR-CNN and I guess the author simply followed the name of forward function's variable in cascade_roi_head. Most definitions of RoIHeads use ```x``` to indicate features so it would be a little weird if we change it.